docs/topics/templates.txt

```diff
@@ -415,6 +415,22 @@ adds defaults that differ from Jinja2's for a few options:
 
   It defaults to an empty list.
 
+  Use context processors for non-trivial request-dependent values that you want
+  computed only once per template render (and not at all if no template is
+  rendered).
+
+  If a context processor's value isn't used in a majority of your templates or
+  isn't used more than once per template render, it would be more explicit to
+  put the function that you would use as a context processor in the template
+  context (for example, using ``jinja2.Environment`` as described below) and
+  then to call that function in the template::
+
+    {{ function(request) }}
+
+  Since context processors are run for every template that's rendered, think
+  carefully before putting expensive computations in them, especially if the
+  result isn't used in all or most of your templates.
+
```

Comment:
This proposal looks technically correct to me but I think it could be misinterpreted. 

The text starts by advising to use context processors while the general goal is to recommend against them. A critical condition is in parentheses.

Here's what I would write -- I know the quality of the wording isn't as good as the proposal above, I focused on the flow of ideas.

----

Using context processors with Jinja2 templates is discouraged.

Context processors are useful with Django templates because Django templates don't support calling functions with arguments. Jinja2 has no such limitation. Therefore it's possible and recommended to put the function that you would use as a context processor in the global variables available to the template (for example, using ``jinja2.Environment`` as described below) and then to call that function in the template::

    {{ function(request) }}

Some built-in context processors for Django templates don't even depend on the request and return a fixed value. For Jinja2 templates, add this value to global variables directly instead of configuring a context processor to add it to the global variables.

Since context processors are run for every template that's rendered, think carefully before putting expensive computations in them, especially if the result isn't used in all or most of your templates.

The only known use case for context processors in Jinja2 templates requires the following conditions:

- making an expensive computation that depends on the request
- needing its results in every template
- using its results multiple times in each template
- not wanting to cache the result in an attribute of the request, for purity

If these conditions aren't met, passing a function to the template in global variables is simpler and more in line with the design of Jinja2.